packages/catalog-server/src/lib/firestore/firestore-repository.ts

```diff
@@ -275,31 +290,60 @@ export class FirestoreRepository implements Repository {
   }
 
   async writeCustomElements(
-    packageName: string,
-    version: string,
+    packageVersionMetadata: Version,
     customElements: CustomElementInfo[],
     distTags: string[],
     author: string
   ): Promise<void> {
     // Store custom elements data in subcollection
+    const {name: packageName, version, description} = packageVersionMetadata;
     const versionRef = this.getPackageVersionRef(packageName, version);
     const customElementsRef = versionRef.collection('customElements');
+    const isLatest = distTags.includes('latest');
     const batch = db.batch();
 
+    // Stem the package description
+    const packageDescriptionStems = natural.PorterStemmer.tokenizeAndStem(
+      description ?? ''
+    );
+
     for (const c of customElements) {
+      const tagName = c.export.name;
+      // Grab longer tag name parts for searching. We want "button" from
+      // md-button, etc.
+      const tagNameParts = tagName.split('-').filter((s) => s.length > 3);
+      const descriptionStems = natural.PorterStemmer.tokenizeAndStem(
+        c.declaration.description ?? ''
+      );
+      const summaryStems = natural.PorterStemmer.tokenizeAndStem(
+        c.declaration.summary ?? ''
+      );
+
+      // Combine and deduplicate terms
+      const searchTerms = [
+        ...new Set([
+          ...packageDescriptionStems,
+          ...descriptionStems,
+          ...summaryStems,
+          ...tagNameParts,
```

Comment:
Just an intuition, no action needed. Feels like we might want to weight tag name parts a little more highly than description terms?

Reply:
I think weighting / relevance might be one reason to push us to use a search service rather quickly. I don't think we have a natural way of getting Firestore to consider some kind of weighting in our query, we just get a result or not for the `array-contains-any` operator. Sorting by array fields [isn't very useful](https://cloud.google.com/firestore/docs/concepts/data-types). We can get Firestore to sort by simple fields like a "ranking" field.

To do a relevance ordering we'd have to talk all the results, iterate and apply something like tf-id, and sort on that. That means our Firestore results size will have to be bigger than our requested query size because  we might have high-relevance results later in the Firestore result set.

It's plausible we could continue down this path and keep it simple, but it also seems like it might get complicated real quick.

Reply:
Good point,it  didn't occur to me that we'd need the ranking to happen on the database side, and presumably you couldn't write a query like that (pretty sure you could with something like postgres, but firebase doubtful). I'm pretty convinced we'll need a standalone search service, then, based on lucene or similar (a service we run, or a third-party hosted one).